common/src/main/java/com/microsoft/identity/common/crypto/AndroidBrokerStorageEncryptionManager.java

```diff
@@ -0,0 +1,173 @@
+                                                 @Nullable final ITelemetryCallback telemetryCallback) {
+        mContext = context;
+        mTelemetryCallback = telemetryCallback;
+
+        mLegacyAuthAppKeyLoader = new PredefinedKeyLoader(LEGACY_AUTHENTICATOR_APP_KEY_ALIAS,
+                AuthenticationSettings.INSTANCE.getBrokerSecretKeys().get(AZURE_AUTHENTICATOR_APP_PACKAGE_NAME));
+
+        mLegacyCPKeyLoader = new PredefinedKeyLoader(LEGACY_COMPANY_PORTAL_KEY_ALIAS,
+                AuthenticationSettings.INSTANCE.getBrokerSecretKeys().get(COMPANY_PORTAL_APP_PACKAGE_NAME));
+
+        mKeyStoreKeyLoader = new AndroidWrappedKeyLoader(KEY_STORE_ALIAS, context, telemetryCallback);
+    }
+
+    // Exposed for Robolectric.
+    protected String getPackageName(){
+        return mContext.getPackageName();
+    }
+
+    @Override
+    public @NonNull AbstractSecretKeyLoader getKeyLoaderForEncryption() {
+        if (StorageHelper.sShouldEncryptWithKeyStoreKey) {
+            return mKeyStoreKeyLoader;
+        }
+
+        if (AZURE_AUTHENTICATOR_APP_PACKAGE_NAME.equalsIgnoreCase(getPackageName())) {
+            return mLegacyAuthAppKeyLoader;
+        }
+
+        if (COMPANY_PORTAL_APP_PACKAGE_NAME.equalsIgnoreCase(getPackageName()) ||
+                BROKER_HOST_APP_PACKAGE_NAME.equalsIgnoreCase(getPackageName())) {
+            return mLegacyCPKeyLoader;
+        }
+
+        throw new IllegalStateException("Matching encryption key not found");
+    }
+
+    @Override
+    public @NonNull List<AbstractSecretKeyLoader> getKeyLoaderForDecryption(@NonNull byte[] cipherText) {
+        final String methodName = ":getKeyLoaderForDecryption";
+        final String packageName = getPackageName();
+
+        if (isEncryptedByThisKeyIdentifier(cipherText, PredefinedKeyLoader.KEY_IDENTIFIER)) {
+            if (COMPANY_PORTAL_APP_PACKAGE_NAME.equalsIgnoreCase(packageName) ||
+                    BROKER_HOST_APP_PACKAGE_NAME.equalsIgnoreCase(packageName)) {
+                return new ArrayList<AbstractSecretKeyLoader>() {{
+                    add(mLegacyCPKeyLoader);
+                    add(mLegacyAuthAppKeyLoader);
+                }};
+            } else if (AZURE_AUTHENTICATOR_APP_PACKAGE_NAME.equalsIgnoreCase(packageName)) {
+                return new ArrayList<AbstractSecretKeyLoader>() {{
+                    add(mLegacyAuthAppKeyLoader);
+                    add(mLegacyCPKeyLoader);
+                }};
+            } else {
+                throw new IllegalStateException("Unexpected Broker package name. Cannot load key.");
+            }
+        }
+
+        if (isEncryptedByThisKeyIdentifier(cipherText, AndroidWrappedKeyLoader.KEY_IDENTIFIER)) {
+            return new ArrayList<AbstractSecretKeyLoader>() {{
+                add(mKeyStoreKeyLoader);
+            }};
+        }
+
+        Logger.warn(TAG + methodName, "Cannot find a matching key to decrypt the given blob");
+        return new ArrayList<>();
```

Comment:
Are we checking for empty `List` in the result everywhere that this is used?

Reply:
We do. (In StorageEncryptionManager.decrypt())